src/main/java/com/aws/greengrass/componentmanager/DependencyResolver.java

```diff
@@ -106,18 +122,27 @@
                 .log("Start to resolve group dependencies");
         // resolve target components dependencies
         for (String component : targetComponentsToResolve) {
-            resolvedComponents.putAll(resolveComponentDependencies(component, componentNameToVersionConstraints,
+            resolveComponentDependencies(component, componentNameToVersionConstraints,
+                    resolvedComponents, componentIncomingReferenceCount,
                     (name, requirements) -> componentManager.resolveComponentVersion(name, requirements,
-                            document.getDeploymentId())));
+                            document.getDeploymentId()));
+        }
+
+        // detect circular dependencies for target components from the current deployment
```

Comment:
A deployment will not succeed if there is a circular dependency, so we can assume that if are to find a circular dependency it will be in the current deployment.

Reply:
The target components belonging to other deployment/groups has already been checked for circular dependency. If a circular dependency is introduced by a component this deployment brought it (new component or a bump in component version) it would introduce a circular dependency for target components of this deployment as well. Because of this assumption, we only need to check the current deployment/group.